src/client/pages/Offer/ProductSelector/ProductSlector.tsx

```diff
@@ -0,0 +1,159 @@
+  name: string
+  description: string
+  price: string
+  image: string
+}
+
+type ProductSelectorProps = {
+  className?: string
+  mainProducts: Array<Product>
+  additionalProducts: Array<Product>
+}
+
+const getInitialState = ({
+  mainProducts,
+  additionalProducts,
+}: {
+  mainProducts: Array<Product>
+  additionalProducts: Array<Product>
+}): {
+  mainProducts: Record<string, boolean>
+  additionalProducts: Record<string, boolean>
+} => {
+  const getProductIdByValueMap = (products: Array<Product>) =>
+    products.reduce(
+      (result, product) => ({
+        ...result,
+        [product.id]: false,
+      }),
+      {},
+    )
+
+  return {
+    mainProducts: getProductIdByValueMap(mainProducts),
+    additionalProducts: getProductIdByValueMap(additionalProducts),
+  }
+}
+
+export const ProductSelector = ({
+  className,
+  mainProducts,
+  additionalProducts,
+}: ProductSelectorProps) => {
+  const textKeys = useTextKeys()
+  const inputRef = useRef<HTMLInputElement | null>(null)
+  const [selectedProductsByCategory, setSelectedProductsByCategory] = useState(
+    getInitialState({ mainProducts, additionalProducts }),
+  )
+
+  const selectedMainProducts = useMemo(() => {
+    return Object.keys(selectedProductsByCategory.mainProducts).filter(
+      (productId) => selectedProductsByCategory.mainProducts[productId],
+    )
+  }, [selectedProductsByCategory])
+
+  return (
+    <Wrapper className={className}>
+      <section>
+        <h2>Main Coverage</h2>
+        <MainCoverageCardGrid>
+          {mainProducts.map(({ id, name, price, description, image }) => {
+            const isTheOnlySelectedMainProduct =
+              selectedMainProducts.length === 1 &&
+              selectedMainProducts[0] === id
+
+            return (
+              <MainProductCard
+                key={id}
+                checkboxRef={
+                  isTheOnlySelectedMainProduct ? inputRef : undefined
+                }
+                title={name}
+                price={price}
+                description={description}
+                image={image}
+                checked={selectedProductsByCategory.mainProducts[id]}
+                onClick={() => {
+                  if (isTheOnlySelectedMainProduct) {
+                    inputRef.current?.setCustomValidity(
+                      textKeys.OFFER_PAGE_MISSING_MAIN_COVERAGE_ERROR(),
+                    )
+                    inputRef.current?.reportValidity()
+                  } else {
+                    setSelectedProductsByCategory((prevState) => ({
+                      ...prevState,
+                      mainProducts: {
+                        ...prevState.mainProducts,
+                        [id]: !prevState.mainProducts[id],
+                      },
+                    }))
+                  }
+                }}
+              />
+            )
+          })}
+        </MainCoverageCardGrid>
+      </section>
+      <section>
+        <h2>Additional Coverage</h2>
```

Comment:
Are we adding translated keys already?

Reply:
Good point. I guess we can already add some text keys for those headings. Will do.